### PR TITLE
Nth Root

### PR DESCRIPTION
**Pull request by aliang8**
Nth Root rules and test cases

ChangeTypes from `mathsteps`
https://github.com/socraticorg/mathsteps/blob/master/lib/ChangeTypes.js#L128-L148

### CODE REVIEW

File: test/rules_test.js [hunk 1]
```diff
@@ -217,6 +217,61 @@ describe('rules', () => {
         ['x^(|-(x + 1)|)', 'x^(x + 1)'],
     ])
 
+    suite('simplify fraction', rules.SIMPLIFY_FRACTION, [
+        ['1/3', '1 / 3'],
+        ['2/6', '1 / 3'],
+        ['15/24', '5 / 8']
+    ])
```
Comment (k4b7) [inline test/rules_test.js]: What happens with negative numbers or if the numerator is `0`?

File: lib/rules.js [hunk 2]
```diff
@@ -184,6 +185,234 @@ export const MULTIPLY_BY_INVERSE = defineRuleString('#a / (#b / #c)', '#a * (#c
...
+// assumes that the exponent is a constant value
+// TODO: handle cases like nthRoot('x^2y', 2)
+export const CANCEL_EXPONENT = defineRule(
+    (node) => {
+        let isNthRoot = false
+        let validRadicand = false
+        if (query.isApply(node)){
+            isNthRoot = query.isIdentifier(node.op)
+                && node.op.name == 'nthRoot'
+            validRadicand = query.isPow(node.args[0])
```
Comment (k4b7) [inline lib/rules.js]: I think `nthRoot(x, 2)` should probably be parsed as
```
{ 
  type: 'Apply',
  op: 'nthRoot',
  args: [ 
    { type: 'Identifier', name: 'x' },
    { type: 'Number', value: '2' },
  ],
}
```
to differentiate from being a function.  The benefit being that when an operator is an identifier, we can always render it as a function without having to do additional logic.  I'll update the parser.

File: lib/rules.js [hunk 3]
```diff
@@ -184,6 +185,234 @@ export const MULTIPLY_BY_INVERSE = defineRuleString('#a / (#b / #c)', '#a * (#c
...
+        // Case #1: numerator > denominator && denominator = 1
+        // e.g nthRoot(x^4, 2) -> 4/2 -> 2/1 -> x^2
+        // Case #2: numerator < denominator ||
+        // numerator > denominator && denominator != 1, return nthRoot
+        // e.g nthRoot(x^2, 4) -> 2/4 -> 1/2 -> nthRoot(x^1, 2)
+        // Case #3: numerator == denominator
+        // e.g nthRoot(x^2, 2) -> 2/2 -> x^1
+
+        if (query.getValue(numerator) > query.getValue(denominator)
+            &&  query.getValue(denominator) == 1){
```
Comment (k4b7) [inline lib/rules.js]: Single space after `&&` and between `) {`.

File: lib/rules.js [hunk 4]
```diff
@@ -184,6 +185,234 @@ export const MULTIPLY_BY_INVERSE = defineRuleString('#a / (#b / #c)', '#a * (#c
...
+            result = build.applyNode('pow', [variable, numerator])
+        } else if (query.getValue(numerator) === query.getValue(denominator)){
+            result = build.applyNode('pow', [variable, build.numberNode(1)])
+        } else {
+            result = build.applyNode (
+                {type: 'Identifier', name: 'nthRoot'},
+                [build.applyNode('pow', [variable, numerator]), denominator]
+            )
```
Comment (k4b7) [inline lib/rules.js]: These values could be returned removing the need for a `result` variable.

File: lib/rules.js [hunk 5]
```diff
@@ -184,6 +185,234 @@ export const MULTIPLY_BY_INVERSE = defineRuleString('#a / (#b / #c)', '#a * (#c
...
+        let numerator = newRoot.args[0]
+        let denominator = newRoot.args[1]
+
+        // Case #1: numerator > denominator && denominator = 1
+        // e.g nthRoot(x^4, 2) -> 4/2 -> 2/1 -> x^2
+        // Case #2: numerator < denominator ||
+        // numerator > denominator && denominator != 1, return nthRoot
+        // e.g nthRoot(x^2, 4) -> 2/4 -> 1/2 -> nthRoot(x^1, 2)
+        // Case #3: numerator == denominator
+        // e.g nthRoot(x^2, 2) -> 2/2 -> x^1
```
Comment (k4b7) [inline lib/rules.js]: These comments don't seem to line up with the cases below... it seems like cases 2 and 3 should be flipped.

File: lib/rules.js [hunk 6]
```diff
@@ -184,6 +185,234 @@ export const MULTIPLY_BY_INVERSE = defineRuleString('#a / (#b / #c)', '#a * (#c
...
+                {type: 'Identifier', name: 'nthRoot'},
+                [build.applyNode('pow', [variable, numerator]), denominator]
+            )
+        }
+        return result
+    }
+)
+
+// e.g. nthRoot(2, 2) * nthRoot(3, 2) -> nthRoot(2 * 3, 2)
+export const COMBINE_UNDER_ROOT = defineRuleString('nthRoot(#a, #c) * nthRoot(#b, #c)', 'nthRoot(#a * #b, #c)')
```
Comment (k4b7) [inline lib/rules.js]: Cool!  Can we make this n-ary aware?

File: lib/rules.js [hunk 7]
```diff
@@ -184,6 +185,234 @@ export const MULTIPLY_BY_INVERSE = defineRuleString('#a / (#b / #c)', '#a * (#c
...
+
+// e.g. 2 * 2 * 2 -> 2 ^ 3
+export const CONVERT_MULTIPLICATION_TO_EXPONENT = defineRuleString('#a^#b_0 * ...', '#a^#eval(#b_0 + ...)', {a: query.isNumber, b: query.isNumber})
```
Comment (k4b7) [inline lib/rules.js]: This pattern doesn't match the example.  You could try:
```
'#a * ...', '#a^#eval(#a^0 + ...)'
```
I don't know if it will work.  If it doesn't that's something we should fix.  Tracking this in https://github.com/semantic-math/math-rules/issues/28.

File: lib/rules.js [hunk 8]
```diff
@@ -184,6 +185,234 @@ export const MULTIPLY_BY_INVERSE = defineRuleString('#a / (#b / #c)', '#a * (#c
...
+
+// e.g. nthRoot(2 * x) -> nthRoot(2) * nthRoot(x)
+export const DISTRIBUTE_NTH_ROOT = defineRule(
```
Comment (k4b7) [inline lib/rules.js]: I think this can be written as a pattern:
```
nthRoot(#a_0 * ...) --> nthRoot(#a_0) * ...
```
If that doesn't work, please file a bug because we should support this kind of pattern.

File: lib/rules.js [hunk 9]
```diff
@@ -184,6 +185,234 @@ export const MULTIPLY_BY_INVERSE = defineRuleString('#a / (#b / #c)', '#a * (#c
...
+        )
+        return result
+    }
+)
+
+// e.g. nthRoot(4) * nthRoot(x^2) -> 2 * x
+export const EVALUATE_DISTRIBUTED_NTH_ROOT = defineRule(
+    (node) => {
+        let isDistributed = query.isMul(node)
+        let canEvaluate = canApplyRule(CANCEL_EXPONENT, node)
```
Comment (k4b7) [inline lib/rules.js]: This will fail for `4`, unless I'm misunderstanding what `CANCEL_EXPONENT` does.

File: lib/rules.js [hunk 10]
```diff
@@ -184,6 +185,234 @@ export const MULTIPLY_BY_INVERSE = defineRuleString('#a / (#b / #c)', '#a * (#c
...
+
+                    const args = []
+
+                    while(leftover - times > 0) {
+                        leftover -= times
+                        args.push(Array(times).fill(build.numberNode(key)))
+                    }
+                    leftover == 1
+                        ? args.push(build.numberNode(key))
+                        : args.push(Array(leftover).fill(build.numberNode(key)))
```
Comment (k4b7) [inline lib/rules.js]: I think this last line should be
```
: args.push(...Array(leftover).fill(build.numberNode(key)))
```
It should be possible to remove the ternary and use:
```
args.push(...Array(leftover).fill(build.numberNode(key)))
```

File: lib/rules.js [hunk 11]
```diff
@@ -184,6 +185,234 @@ export const MULTIPLY_BY_INVERSE = defineRuleString('#a / (#b / #c)', '#a * (#c
...
+
+                    args.map(arg => {
+                        if (Array.isArray(arg)) {
+                            const loc = args.indexOf(arg)
+                            args[loc] = build.applyNode('mul', arg)
+                        }
+                    })
+
+                    return args
```
Comment (k4b7) [inline lib/rules.js]: Don't use `map` unless you use the return value.  Also, you can get the index of the item as it's one of the callback args.
```
args.forEach((arg, index) => {
   if (Array.isArray(arg)) {
      args[index] = build.applyNode('mul', arg)
   }
}
```

I would have a slight preference for `map` used in the following way:
```
return args.map(arg =>  Array.isArray(arg)
   ? build.applyNode('mul', arg)
   : arg)
```

File: lib/rules.js [hunk 12]
```diff
@@ -184,6 +185,234 @@ export const MULTIPLY_BY_INVERSE = defineRuleString('#a / (#b / #c)', '#a * (#c
...
+            {type: 'Identifier', name: 'nthRoot'},
+            [build.applyNode(
+                'mul',
+                flatten(Object.keys(count).map(key => {
+                    let leftover = count[key]
+                    const times = query.getValue(index)
+
+                    const args = []
+
+                    while(leftover - times > 0) {
```
Comment (k4b7) [inline lib/rules.js]: space between `while` and `(`

File: lib/rules.js [hunk 13]
```diff
@@ -184,6 +185,234 @@ export const MULTIPLY_BY_INVERSE = defineRuleString('#a / (#b / #c)', '#a * (#c
...
+                    [arg]
+                )
+            })
+        )
+        return result
+    }
+)
+
+// e.g. nthRoot(4) * nthRoot(x^2) -> 2 * x
+export const EVALUATE_DISTRIBUTED_NTH_ROOT = defineRule(
```
Comment (k4b7) [inline lib/rules.js]: If we solve https://github.com/semantic-math/math-rules/issues/23 then rules like this should be trivial.  It'd be great to solve it in a general way so that we don't have to write custom functions each time we want to apply the same rule to multiple nodes.  In this case we're applying one rule or another so we'd nee to https://github.com/semantic-math/math-rules/issues/10 as well.

File: lib/rules.js [hunk 14]
```diff
@@ -184,6 +185,234 @@ export const MULTIPLY_BY_INVERSE = defineRuleString('#a / (#b / #c)', '#a * (#c
...
+                                     {type: 'Identifier', name: 'nthRoot'},
+                                     [nthRoot.args[0]]
+                                 ))
+            })
+        )
+        return result
+    }
+)
+
+export const primeFactorization = (number, result = []) => {
```
Comment (k4b7) [inline lib/rules.js]: Would be nice to reuse https://github.com/semantic-math/math-evaluator/blob/master/lib/functions.js#L1-L20.

File: lib/rules.js [hunk 15]
```diff
@@ -184,6 +185,234 @@ export const MULTIPLY_BY_INVERSE = defineRuleString('#a / (#b / #c)', '#a * (#c
...
+        return query.isNumber(node) ? {node} : null
+    },
+
+    (node) => {
+        const factors = primeFactorization(query.getValue(node))
+        const result = build.applyNode(
+            'mul',
+            factors.map(arg => {
+                return build.numberNode(arg)
+            })
```
Comment (k4b7) [inline lib/rules.js]:
```
factors.map(build.numberNode),
```

File: lib/rules.js [hunk 16]
```diff
@@ -184,6 +185,234 @@ export const MULTIPLY_BY_INVERSE = defineRuleString('#a / (#b / #c)', '#a * (#c
...
+                }))
+            ), index]
+        )
+
+        return result
+    }
+)
+
+// e.g nthRoot(4) -> 2
+export const NTH_ROOT_VALUE = defineRuleString('nthRoot(#a)', '#eval(#a^(1/2))', {a: query.isNumber})
```
Comment (k4b7) [inline lib/rules.js]: This should handle all `nthRoots` not just square roots.  We can enforce that `nthRoot` nodes always have two args in the parser so that we always have an index.  That way we don't have to have special case logic to handle the case where it doesn't exist.

File: test/rules_test.js [hunk 17]
```diff
@@ -217,6 +217,61 @@ describe('rules', () => {
...
+
+    suite('cancel exponent', rules.CANCEL_EXPONENT, [
+        ['nthRoot(x^2, 4)', 'nthRoot(x^1, 2)'],
+        ['nthRoot(y^3, 4)', 'nthRoot(y^3, 4)'],
+        ['nthRoot(a^15, 24)', 'nthRoot(a^5, 8)'],
+        ['nthRoot(b^4, 2)', 'b^2'],
+        ['nthRoot(c^8, 3)', 'nthRoot(c^8, 3)'],
+        ['nthRoot(d^10, 10)', 'd^1'],
+        ['nthRoot(x^2)', 'x^1'],
```
Comment (k4b7) [inline test/rules_test.js:236]: Can you test cases with negative exponents and negative indexes?

File: test/rules_test.js [hunk 18]
```diff
@@ -217,6 +217,61 @@ describe('rules', () => {
...
+    ])
+
+    suite('combine under root', rules.COMBINE_UNDER_ROOT, [
+        ['nthRoot(2, 2) * nthRoot(3, 2)', 'nthRoot(2 * 3, 2)'],
+        ['nthRoot(4, 3) * nthRoot(3, 3)', 'nthRoot(4 * 3, 3)']
+    ])
```
Comment (k4b7) [inline test/rules_test.js]: Can you add tests cases for n-ary instances as well as cases where the radicand and the index are identifiers?

File: test/rules_test.js [hunk 19]
```diff
@@ -217,6 +217,61 @@ describe('rules', () => {
...
+
+    suite('distribute nthRoot', rules.DISTRIBUTE_NTH_ROOT, [
+        ['nthRoot(2 * x)', 'nthRoot(2) * nthRoot(x)'],
+        ['nthRoot(3 * 3 * x)', 'nthRoot(3) * nthRoot(3) * nthRoot(x)']
+    ])
+
+    suite('convert multiplication to exponent', rules.CONVERT_MULTIPLICATION_TO_EXPONENT, [
+        ['2^1 * 2^1 * 2^3', '2^5'],
+        ['3^2 * 3^1 * 3^20', '3^23']
```
Comment (k4b7) [inline test/rules_test.js]: Interesting... so this means that you have to add an exponent to a base before running this rule.  That makes sense.  Under what circumstances do we run the ADD_ONE_EXPONENT rule?

File: test/rules_test.js [hunk 20]
```diff
@@ -217,6 +217,61 @@ describe('rules', () => {
...
+    ])
+
+    suite('evaluate distributed nthRoot', rules.EVALUATE_DISTRIBUTED_NTH_ROOT, [
+        ['nthRoot(4) * nthRoot(x^2)', '2 * x^1'],
+        ['nthRoot(x^3) * nthRoot(36)', 'nthRoot(x^3, 2) * 6']
```
Comment (k4b7) [inline test/rules_test.js]: Can you add test cases where the index is not `2`?

File: test/rules_test.js [hunk 21]
```diff
@@ -217,6 +217,61 @@ describe('rules', () => {
...
+    ])
+
+    suite('factor into prime', rules.FACTOR_INTO_PRIME, [
+        ['12' ,'2 * 2 * 3'],
+        ['36', '2 * 2 * 3 * 3'],
+        ['91', '7 * 13'],
+        ['2', '2'],
+        ['1', '1']
```
Comment (k4b7) [inline test/rules_test.js]: What happens with negative numbers?

File: test/rules_test.js [hunk 22]
```diff
@@ -217,6 +217,61 @@ describe('rules', () => {
...
+    ])
+
+    suite('group terms by root', rules.GROUP_TERMS_BY_ROOT, [
+        ['nthRoot(2 * 2 * 2 * 3, 2)', 'nthRoot((2 * 2) * 2 * 3, 2)'],
+        ['nthRoot(2 * 3 * 3 * 2, 3)', 'nthRoot((2 * 2) * (3 * 3), 3)'],
+        ['nthRoot(5 * 7 * 9 * 7 * 7 * 7, 4)', 'nthRoot(5 * (7 * 7 * 7 * 7) * 9, 4)']
```
Comment (k4b7) [inline test/rules_test.js]: Can you add tests for negative numbers, identifiers, and factors w/ exponents?

File: test/rules_test.js [hunk 23]
```diff
@@ -217,6 +217,61 @@ describe('rules', () => {
...
+    ])
+
+    suite('nthRoot value', rules.NTH_ROOT_VALUE, [
+        ['nthRoot(4)', '2'],
+        ['nthRoot(16)', '4']
```
Comment (k4b7) [inline test/rules_test.js]: Can you add tests for other indexes?  And negative values for odd indexes?

File: test/rules_test.js [hunk 24]
```diff
@@ -349,6 +404,11 @@ describe('rules', () => {
         ['-(a + b + c)', '-1 * a + -1 * b + -1 * c'],
     ])
 
+    suite('cancel exponent', rules.NTH_ROOT_VALUE, [
+        ['nthRoot(4)', '2'],
+        ['nthRoot(16)', '4'],
+        ['nthRoot(64)', '8']
+    ])
```
Comment (k4b7) [inline test/rules_test.js]: Duplicate test suite.

**Review by k4b7 — CHANGES_REQUESTED** [review]

File: lib/rules.js (hunk 14 above)
Comment (aliang8) [reply lib/rules.js]: I need it to be exposed :D

File: lib/rules.js (hunk 10 above)
Comment (aliang8) [reply lib/rules.js]: It doesn't preserve the parentheses if I do that:
```
nthRoot(2 * 2 * 3 * 3, 3)
```
vs
```
nthRoot((2 * 2) * (3 * 3), 3)
```

File: lib/rules.js (hunk 11 above)
Comment (aliang8) [reply lib/rules.js]: Thanks! Will keep that in mind.

File: lib/rules.js (hunk 5 above)
Comment (aliang8) [reply lib/rules.js]: Fixed

File: lib/rules.js (hunk 3 above)
Comment (aliang8) [reply lib/rules.js]: Fixed

File: lib/rules.js (hunk 4 above)
Comment (aliang8) [reply lib/rules.js]: Fixed

File: lib/rules.js (hunk 7 above)
Comment (aliang8) [reply lib/rules.js]: The example is misleading. Sorry about that. A better example would be `2^1 * 2^2 * 2^3` -> `2^6`

File: lib/rules.js (hunk 8 above)
Comment (aliang8) [reply lib/rules.js]: Filed issue

File: lib/rules.js (hunk 15 above)
Comment (aliang8) [reply lib/rules.js]: Fixed

File: lib/rules.js (hunk 12 above)
Comment (aliang8) [reply lib/rules.js]: Fixed

File: test/rules_test.js (hunk 1 above)
Comment (aliang8) [reply test/rules_test.js]: Fixed

File: test/rules_test.js (hunk 21 above)
Comment (aliang8) [reply test/rules_test.js]: I'm not exactly sure how I can handle negative numbers. In mathematics, negative numbers do not have prime factorizations. I don't see how it is possible for me to match negative numbers without matching the positive version. 

```
{ type: 'Number', value: '12' }
{ type: 'Apply',
  op: 'neg',
  args: [ { type: 'Number', value: '12' } ] }
```
As you can see, `12` comes before `-12` in the `matchPattern`.

File: test/rules_test.js (hunk 24 above)
Comment (aliang8) [reply test/rules_test.js]: Fixed

File: test/rules_test.js (hunk 23 above)
Comment (aliang8) [reply test/rules_test.js]: Fixed. Not sure why it doesn't work for negative bases, but it works for negative indices.

File: test/rules_test.js (hunk 17 above)
Comment (aliang8) [reply test/rules_test.js:236]: Fixed

File: test/rules_test.js (hunk 18 above)
Comment (aliang8) [reply test/rules_test.js]: Fixed

File: lib/rules.js (hunk 6 above)
Comment (aliang8) [reply lib/rules.js]: Fixed

File: lib/rules.js [hunk 25]
```diff
@@ -146,7 +147,49 @@ export const CANCEL_MINUSES = defineRuleString('-#a / -#b', '#a / #b')
 // CANCEL_TERMS: 'CANCEL_TERMS',
 
 // e.g. 2/6 -> 1/3
-// SIMPLIFY_FRACTION: 'SIMPLIFY_FRACTION',
+/*
+export const SIMPLIFY_FRACTION =
+    defineRuleString('#a / #b', '(#eval(#a/gcd(#a, #b)) / (#eval(#b/gcd(#a, #b))))', {a: query.isNumber, b: query.isNumber})
+*/
+export const SIMPLIFY_FRACTION = defineRule(
```
Comment (k4b7) [inline lib/rules.js]: I've update `math-evaluator`'s `gcd` to handle negatives and adding support for evaluating `abs` nodes.  The following works now:
```
export const SIMPLIFY_FRACTION = defineRuleString(
    '#a / #b',
    '#eval(sign(#a/#b) * |#a|/gcd(#a, #b)) / #eval(|#b|/gcd(#a, #b))',
    {a: query.isNumber, b: query.isNumber})
```
The only case it doesn't handle which is `0 / 3` which is already handled by `REDUCE_ZERO_NUMERATOR`.  I think shouldn't worry about this other case for now.  If we do need to handle as part of `SIMPLIFY_FRACTION`, https://github.com/semantic-math/math-rules/issues/10 should make it easy to combine rules.

File: lib/rules.js [hunk 26]
```diff
@@ -184,6 +227,185 @@ export const MULTIPLY_BY_INVERSE = defineRuleString('#a / (#b / #c)', '#a * (#c
...
+        let denomVal = query.getValue(denominator)
+
+        // Case #1: numerator > denominator || numerator == -1 && denominator = 1
+        // e.g nthRoot(x^4, 2) -> 4/2 -> 2/1 -> x^2
+        // Case #2: numerator == denominator
+        // e.g nthRoot(x^2, 2) -> 2/2 -> x^1
+        // Case #3: numerator < denominator ||
+        // numerator > denominator && denominator != 1, return nthRoot
+        // e.g nthRoot(x^2, 4) -> 2/4 -> 1/2 -> nthRoot(x^1, 2)
+        if ((numVal > denomVal || numVal == -1) && denomVal == 1) {
```
Comment (k4b7) [inline lib/rules.js]: The condition doesn't match the condition in the comment, the parens are in the wrong spots.

File: lib/rules.js [hunk 27]
```diff
@@ -184,6 +227,185 @@ export const MULTIPLY_BY_INVERSE = defineRuleString('#a / (#b / #c)', '#a * (#c
...
+
+        // simplify exponent / index
+        // e.g. nthRoot(x^2, 4) -> 2/4 -> 1/2
+        const newRoot = applyRule(SIMPLIFY_FRACTION,
+                                  build.applyNode('div', [exponent, index]))
+
+        let numerator = newRoot.args[0]
+        let denominator = newRoot.args[1]
+        let numVal = query.getValue(numerator)
+        let denomVal = query.getValue(denominator)
```
Comment (k4b7) [inline lib/rules.js]: Calling these `numerator` and `denominator` is a bit misleading.  Can you rename these to `exponent` and `index`?

File: lib/rules.js [hunk 28]
```diff
@@ -184,6 +227,185 @@ export const MULTIPLY_BY_INVERSE = defineRuleString('#a / (#b / #c)', '#a * (#c
...
+// e.g. nthRoot(x^2, 4) -> nthRoot(x^1 ,2)
+export const CANCEL_EXPONENT = defineRule(
+    (node) => {
+        let isNthRoot = false
+        let validRadicand = false
+        if (query.isApply(node)) {
+            isNthRoot = (node.op == 'nthRoot')
+            validRadicand = query.isPow(node.args[0])
+        }
+        return (isNthRoot && validRadicand) ? {node} : null
```
Comment (k4b7) [inline lib/rules.js:204]: Is there an assumption that the exponent and index are both positive?  If so, we should return false when those criteria are not met.

File: lib/rules.js [hunk 29]
```diff
@@ -184,6 +227,185 @@ export const MULTIPLY_BY_INVERSE = defineRuleString('#a / (#b / #c)', '#a * (#c
...
+// TODO: #10 and #23 (`or` rules and applying multiple rules if
+// multiple occurrences)
+// e.g. nthRoot(4, 2) * nthRoot(x^2, 2) -> 2 * x
+export const EVALUATE_DISTRIBUTED_NTH_ROOT = defineRule(
+    (node) => {
+        let isDistributed = query.isMul(node)
+        let canEvaluate = false
+        if (isDistributed) {
+            canEvaluate = node.args.every(arg => canApplyRule(CANCEL_EXPONENT, arg) || canApplyRule(NTH_ROOT_VALUE, arg))
+        }
```
Comment (k4b7) [inline lib/rules.js:261]: This will fail on things like `x * nthRoot(4, 2) * nthRoot(x^2, 2) * y`.  Please add a test case for this but comment it out.  We can enable it once we have #10.  It would be nice if there was a way to apply a rule from within a pattern.  I've opened an issue for this https://github.com/semantic-math/math-rules/issues/39.

File: lib/rules.js [hunk 30]
```diff
@@ -184,6 +227,185 @@ export const MULTIPLY_BY_INVERSE = defineRuleString('#a / (#b / #c)', '#a * (#c
...
+        return (isDistributed && canEvaluate) ? {node} : null
+    },
+
+    (node) => {
+        const result = build.applyNode(
+            'mul',
+            node.args.map(nthRoot => {
+                let radicand = nthRoot.args[0]
+                let index = nthRoot.args[1]
+                console.log(index)
```
Comment (k4b7) [inline lib/rules.js]: remove

File: lib/rules.js [hunk 31]
```diff
@@ -184,6 +227,185 @@ export const MULTIPLY_BY_INVERSE = defineRuleString('#a / (#b / #c)', '#a * (#c
...
+        }
+        return (isDistributed && canEvaluate) ? {node} : null
+    },
+
+    (node) => {
+        const result = build.applyNode(
+            'mul',
+            node.args.map(nthRoot => {
+                let radicand = nthRoot.args[0]
+                let index = nthRoot.args[1]
```
Comment (k4b7) [inline lib/rules.js]:
```
const [radicand, index] = nthRoot.args
```

File: lib/rules.js [hunk 32]
```diff
@@ -184,6 +227,185 @@ export const MULTIPLY_BY_INVERSE = defineRuleString('#a / (#b / #c)', '#a * (#c
...
+        return query.isNumber(node) ? {node} : null
+    },
+
+    (node) => {
+        const factors = primeFactorization(query.getValue(node))
+        const result = build.applyNode(
+            'mul',
+            factors.map(build.numberNode)
+        )
+        return result
```
Comment (k4b7) [inline lib/rules.js]: No need for a `result` var here,
```
return build.applyNode('mul', factors.map(build.numberNode))
```

File: lib/rules.js [hunk 33]
```diff
@@ -184,6 +227,185 @@ export const MULTIPLY_BY_INVERSE = defineRuleString('#a / (#b / #c)', '#a * (#c
...
+    }
+)
+
+// e.g. nthRoot(2 * 2 * 2, 2) -> nthRoot((2 * 2) * 2)
```
Comment (k4b7) [inline lib/rules.js]: The output should also have an `index`.

File: lib/rules.js [hunk 34]
```diff
@@ -184,6 +227,185 @@ export const MULTIPLY_BY_INVERSE = defineRuleString('#a / (#b / #c)', '#a * (#c
...
+export const GROUP_TERMS_BY_ROOT = defineRule(
+    (node) => {
+        let canGroupTerms = false
+        canGroupTerms = canApplyRule(DISTRIBUTE_NTH_ROOT, node)
```
Comment (k4b7) [inline lib/rules.js]:
```
const canGroupTerms = ...
```

File: lib/rules.js [hunk 35]
```diff
@@ -184,6 +227,185 @@ export const MULTIPLY_BY_INVERSE = defineRuleString('#a / (#b / #c)', '#a * (#c
...
+
+    (node) => {
+        const radicand = node.args[0]
+        const index = node.args[1]
+
+        // dictionary storing the number of times a constant appears
+        // e.g. 2 * 2 * 2 => {'2': 3} , 2 appears 3 times
+        const count = {}
+        radicand.args.forEach(arg => {
+            let key = JSON.stringify(arg)
```
Comment (k4b7) [inline lib/rules.js]:
```
const key = ...
```

File: lib/rules.js [hunk 36]
```diff
@@ -184,6 +227,185 @@ export const MULTIPLY_BY_INVERSE = defineRuleString('#a / (#b / #c)', '#a * (#c
...
+            []
+        )
+
+        const result = build.applyNode(
+            'nthRoot',
+            [build.applyNode(
+                'mul',
+                flatten(Object.keys(count).map(key => {
+                    let leftover = count[key]
+                    let term = JSON.parse(key)
```
Comment (k4b7) [inline lib/rules.js]: const

File: lib/rules.js [hunk 37]
```diff
@@ -184,6 +227,185 @@ export const MULTIPLY_BY_INVERSE = defineRuleString('#a / (#b / #c)', '#a * (#c
...
+            let key = JSON.stringify(arg)
+            count[key] ? count[key]++ : count[key] = 1
+        })
+
+        const flatten = arr => arr.reduce(
+            (acc, val) => acc.concat(
+                Array.isArray(val) ? flatten(val) : val
+            ),
+            []
+        )
```
Comment (k4b7) [inline lib/rules.js]: I think `val` is always an `array` in this case so this can be simplified.

File: lib/rules.js [hunk 38]
```diff
@@ -184,6 +227,185 @@ export const MULTIPLY_BY_INVERSE = defineRuleString('#a / (#b / #c)', '#a * (#c
...
+
+                    const args = []
+
+                    while (leftover - times > 0) {
+                        leftover -= times
+                        args.push(Array(times).fill(term))
+                    }
+                    leftover == 1
+                        ? args.push(term)
+                        : args.push(Array(leftover).fill(term))
```
Comment (k4b7) [inline lib/rules.js]: Can you change this to an `if-else` statement?  `ternaries` that rely on side-effects are confusing.

Comment (k4b7) [reply lib/rules.js]: You might want to reformulate this as:
```
const arg = leftover === 1
   ? term
   : build.applyNode('mul', Array(leftover).fill(term))
args.push(arg)
```
and then skip the `args.forEach` below.

**Review by k4b7 — CHANGES_REQUESTED** [review]

File: lib/rules.js (hunk 25 above)
Comment (aliang8) [reply lib/rules.js]: Fixed

File: lib/rules.js (hunk 27 above)
Comment (aliang8) [reply lib/rules.js]: Fixed

File: lib/rules.js (hunk 26 above)
Comment (aliang8) [reply lib/rules.js]: Fixed

File: lib/rules.js (hunk 29 above)
Comment (aliang8) [reply lib/rules.js:261]: Fixed

File: lib/rules.js (hunk 31 above)
Comment (aliang8) [reply lib/rules.js]: Fixed

File: lib/rules.js (hunk 30 above)
Comment (aliang8) [reply lib/rules.js]: Fixed

File: lib/rules.js (hunk 32 above)
Comment (aliang8) [reply lib/rules.js]: Fixed

File: lib/rules.js (hunk 33 above)
Comment (aliang8) [reply lib/rules.js]: Fixed

File: lib/rules.js (hunk 34 above)
Comment (aliang8) [reply lib/rules.js]: Fixed

File: lib/rules.js (hunk 35 above)
Comment (aliang8) [reply lib/rules.js]: Fixed

File: lib/rules.js (hunk 37 above)
Comment (aliang8) [reply lib/rules.js]:
``` 
(acc, val) => acc.concat(
                flatten(val)
            )
```
doesn't work

File: lib/rules.js (hunk 36 above)
Comment (aliang8) [reply lib/rules.js]: Fixed

File: lib/rules.js (hunk 38 above)
Comment (aliang8) [reply lib/rules.js]: I need the `args.forEach` to generate the parentheses in `nthRoot((2 * 2) * 2 * 3, 2)`, else I would have `nthRoot(2 * 2 * 2 * 3, 2)`

File: lib/rules.js (hunk 28 above)
Comment (aliang8) [reply lib/rules.js:204]: No there is no assumption. It works for any signed combination of exponent and index.

File: lib/rules.js (hunk 38 above)
Comment (aliang8) [reply lib/rules.js]: Not sure why it works for every other case but the one I mentioned.

File: lib/rules.js [hunk 39]
```diff
@@ -253,26 +214,26 @@ export const CANCEL_EXPONENT = defineRule(
...
                                   build.applyNode('div', [exponent, index]))
 
-        let numerator = newRoot.args[0]
-        let denominator = newRoot.args[1]
-        let numVal = query.getValue(numerator)
-        let denomVal = query.getValue(denominator)
+        let newExponent = newRoot.args[0]
+        let newIndex = newRoot.args[1]
+        let numVal = query.getValue(newExponent)
+        let denomVal = query.getValue(newIndex)
```
Comment (k4b7) [inline lib/rules.js]: Can you change these to be `exponentVal` and `indexVal`?

File: lib/rules.js [hunk 40]
```diff
@@ -184,6 +188,179 @@ export const MULTIPLY_BY_INVERSE = defineRuleString('#a / (#b / #c)', '#a * (#c
...
+                    const arg = leftover === 1
+                          ? term
+                          : build.applyNode('mul', Array(leftover).fill(term))
+                    args.push(arg)
+
+                    args.forEach((arg, index) => {
+                        if (Array.isArray(arg)) {
+                            args[index] = build.applyNode('mul', arg)
+                        }
+                    })
```
Comment (kevinbarabash) [inline lib/rules.js]: I don't think this `forEach` is necessary anymore.

File: lib/rules.js [hunk 41]
```diff
@@ -184,6 +188,179 @@ export const MULTIPLY_BY_INVERSE = defineRuleString('#a / (#b / #c)', '#a * (#c
...
+                flatten(Object.keys(count).map(key => {
+                    let leftover = count[key]
+                    const term = JSON.parse(key)
+                    const times = query.getValue(index)
+
+                    const args = []
+
+                    while (leftover - times > 0) {
+                        leftover -= times
+                        args.push(Array(times).fill(term))
```
Comment (k4b7) [inline lib/rules.js]: Instead of adding an array here and then removing it later... we should create the `mul` node right here, e.g.
```
args.push(build.applyNode('mul', Array(times).fill(term)))
```

File: lib/rules.js (hunk 37 above)
Comment (k4b7) [reply lib/rules.js]: Interesting... `Object.keys(count).map(key => { ... }` should be returning an array of arrays since each iteration returns `args` which is an array.  None of the items in `args` should be arrays themselves so flattening shouldn't be necessary.

File: test/rules_test.js (hunk 17 above)
Comment (k4b7) [reply test/rules_test.js:236]: I just realized that I don't have good intuition for what taking taking a root with a negative index means.  I'll have to spend some time thinking about that.  Glad we support it.  :)

File: lib/rules.js (hunk 37 above)
Comment (aliang8) [reply lib/rules.js]: Yeah, I see your point. Not sure why it's failing.

File: lib/rules.js (hunk 41 above)
Comment (aliang8) [reply lib/rules.js]: Ah wonderful! This fix works.

File: lib/rules.js (hunk 39 above)
Comment (aliang8) [reply lib/rules.js]: Fixed

File: lib/rules.js (hunk 37 above)
Comment (aliang8) [reply lib/rules.js]: Fixed

**Review by k4b7 — APPROVED** [review]

**Comment by k4b7** [comment]
Thanks for all the awesome `nthRoot` rules.  🎉
